cpp/src/arrow/util/utf8.cc

```diff
@@ -79,5 +79,31 @@ void InitializeUTF8() {
   std::call_once(utf8_initialized, internal::InitializeLargeTable);
 }
 
+static const uint8_t kBOM[] = "\xef\xbb\xbf";
+
+Status SkipUTF8BOM(const uint8_t* data, int64_t size, const uint8_t** out) {
+  int64_t i;
+  for (i = 0; i < static_cast<int64_t>(sizeof(kBOM) - 1); ++i) {
+    if (size == 0) {
+      if (i == 0) {
+        // Empty string
+        *out = data;
+        return Status::OK();
+      } else {
+        return Status::Invalid("UTF8 string too short (truncated byte order mark?)");
```

Comment:
Should this be an error or should it just avoid skipping it? I have no good enough knowledge of UTF to make such decision.

Reply:
If the streams ends there, it's a truncated UTF8 character, so it's invalid.